src/coreComponents/mesh/MeshObjectPath.hpp

```diff
@@ -0,0 +1,225 @@
+/*
+ * ------------------------------------------------------------------------------------------------------------
+ * SPDX-License-Identifier: LGPL-2.1-only
+ *
+ * Copyright (c) 2018-2020 Lawrence Livermore National Security LLC
+ * Copyright (c) 2018-2020 The Board of Trustees of the Leland Stanford Junior University
+ * Copyright (c) 2018-2020 TotalEnergies
+ * Copyright (c) 2019-     GEOSX Contributors
+ * All rights reserved
+ *
+ * See top level LICENSE, COPYRIGHT, CONTRIBUTORS, NOTICE, and ACKNOWLEDGEMENTS files for details.
+ * ------------------------------------------------------------------------------------------------------------
+ */
+
+/**
+ * @file MeshObjectPath.hpp
+ */
+
+#ifndef GEOSX_MESH_MESHOBJECTPATH_HPP_
+#define GEOSX_MESH_MESHOBJECTPATH_HPP_
+
+
+#include "codingUtilities/EnumStrings.hpp"
+#include "MeshLevel.hpp"
+
+namespace geosx
+{
+class MeshBody;
+
+class MeshObjectPath
+{
+public:
+
+  using permutationMapType = std::map< string, std::map< string, std::map< string, std::vector< string > > > >;
+
+  enum class ObjectTypes : int
+  {
+    nodes,
+    edges,
+    faces,
+    elems,
+    invalid
+  };
+
+
+  MeshObjectPath( string const path,
+                  dataRepository::Group const & meshBodies );
+
+  virtual ~MeshObjectPath();
+
+
+
+  void processPath( string const path,
+                    dataRepository::Group const & meshBodies );
+
+  ObjectTypes const & getObjectType() const
+  {
+    return m_objectType;
+  }
+
+  permutationMapType const & pathPermutations() const
+  {
+    return m_pathPermutations;
+  }
+
+
+  template< typename OBJECT_TYPE = ObjectManagerBase,
+            typename FUNC = std::function< void( dataRepository::Group &) > >
+  void forObjectsInPath( dataRepository::Group & meshBodies,
+                         FUNC && func );
+
+  template< typename OBJECT_TYPE = ObjectManagerBase,
+            typename FUNC = std::function< void( dataRepository::Group const &) > >
+  void forObjectsInPath( dataRepository::Group const & meshBodies,
+                         FUNC && func );
```

Comment:
1. They should both be `const`.
2. It is my impression that I needed a default type for the last template parameter if I use one for the first...perhaps this isn't correct.

Reply:
> 2\. It is my impression that I needed a default type for the last template parameter if I use one for the first...perhaps this isn't correct.

I think it can be omitted if it is derived from an argument. As long as `forObjectsInPath(meshBodies, [](auto){...})` is unambiguous for the compiler.